src/Compilers/CSharp/Portable/Symbols/Source/ExtensionGroupingInfo.cs

```diff
@@ -158,6 +160,53 @@ public Cci.ITypeDefinition GetCorrespondingGroupingType(SourceNamedTypeSymbol ty
             throw ExceptionUtilities.Unreachable();
         }
 
+        internal ImmutableArray<SourceNamedTypeSymbol> GetMatchingExtensions(SourceNamedTypeSymbol extension)
+        {
+            Debug.Assert(extension.IsExtension);
+            return GetCorrespondingMarkerType(extension).UnderlyingExtensions;
+        }
+
+        // Returns all the extension blocks but grouped/merged by equivalency (ie. same marker name)
+        internal IEnumerable<ArrayBuilder<SourceNamedTypeSymbol>> EnumerateMergedExtensionBlocks()
+        {
+            Dictionary<string, MultiDictionary<string, SourceNamedTypeSymbol>> groupingMap = this._groupingMap;
+            ArrayBuilder<string> groupingNames = sortStrings(groupingMap.Keys);
+
+            foreach (var groupName in groupingNames)
+            {
+                MultiDictionary<string, SourceNamedTypeSymbol> markerMap = groupingMap[groupName];
+                ArrayBuilder<string> markerNames = sortStrings(markerMap.Keys);
+
+                foreach (var markerName in markerNames)
+                {
+                    ArrayBuilder<SourceNamedTypeSymbol> extensions = sortSourceNamedTypeSymbols(markerMap[markerName]);
+                    yield return extensions;
+
+                    extensions.Free();
```

Comment:
>extensions.Free(); [](http://example.com/codeflow?start=20&length=18)

This is very fragile, we don't know whether the consumer is done with the builder. Should find other ways to manage builders. Perhaps this method shouldn't allocate new builder (if it is done for sorting, that could be consumer's responsibility), or it should return a builder of builders and the consumer will be responsible to free them once done.

Reply:
Alternatively, consumer can pass a delegate that will be executed for each builder and then the builder will be freed.